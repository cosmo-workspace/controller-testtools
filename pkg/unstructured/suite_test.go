package unstructured

import (
	"os"
	"testing"

	. "github.com/jlandowner/helm-chartsnap/pkg/snap/gomega"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

func TestUnstructured(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Unstructured Suite")
}

var _ = Describe("Diff", func() {
	Context("DiffContextLineN is 3", func() {
		It("should return the extracted diff with previous/next 3 lines", func() {
			expectedSnap := mustReadFile("testdata/expected.snap")
			actualSnap := mustReadFile("testdata/actual.snap")

			d := DiffOptions{
				ContextLineN: 3,
			}
			diff := d.Diff(expectedSnap, actualSnap)
			Ω(diff).To(MatchSnapShot())
		})
	})

	Context("DiffContextLineN is 0", func() {
		It("should return all diff", func() {
			expectedSnap := mustReadFile("testdata/expected.snap")
			actualSnap := mustReadFile("testdata/actual.snap")

			d := DiffOptions{
				ContextLineN: 0,
			}
			diff := d.Diff(expectedSnap, actualSnap)
			Ω(diff).To(MatchSnapShot())
		})
	})
})

var _ = Describe("Unknown", func() {
	Context("OK", func() {
		It("report unknown as warning", func() {
			raw := `some: raw data
raw:
  data: here`
			err := NewUnknownError(raw)

			Ω(err.Error()).To(MatchSnapShot())
		})
	})
})

func mustReadFile(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		panic(err)
	}
	return string(data)
}
