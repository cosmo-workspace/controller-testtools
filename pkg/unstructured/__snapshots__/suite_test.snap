['ApplyDynamicFields should replace specified fields 1']
SnapShot = """
[
  {
    \"apiVersion\": \"apps/v1\",
    \"kind\": \"Deployment\",
    \"metadata\": {
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"###DYNAMIC_FIELD###\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"chartsnap-app1\"
    },
    \"spec\": {
      \"replicas\": 1,
      \"selector\": {
        \"matchLabels\": {
          \"app.kubernetes.io/instance\": \"chartsnap\",
          \"app.kubernetes.io/name\": \"app1\"
        }
      },
      \"template\": {
        \"metadata\": {
          \"labels\": {
            \"app.kubernetes.io/instance\": \"chartsnap\",
            \"app.kubernetes.io/managed-by\": \"Helm\",
            \"app.kubernetes.io/name\": \"app1\",
            \"app.kubernetes.io/version\": \"1.16.0\",
            \"helm.sh/chart\": \"app1-0.1.0\"
          }
        },
        \"spec\": {
          \"containers\": [
            {
              \"image\": \"nginx:1.16.0\",
              \"imagePullPolicy\": \"IfNotPresent\",
              \"livenessProbe\": {
                \"httpGet\": {
                  \"path\": \"/\",
                  \"port\": \"http\"
                }
              },
              \"name\": \"app1\",
              \"ports\": [
                {
                  \"containerPort\": 80,
                  \"name\": \"http\",
                  \"protocol\": \"TCP\"
                }
              ],
              \"readinessProbe\": {
                \"httpGet\": {
                  \"path\": \"/\",
                  \"port\": \"http\"
                }
              },
              \"resources\": {},
              \"securityContext\": {}
            }
          ],
          \"securityContext\": {},
          \"serviceAccountName\": \"IyMjRFlOQU1JQ19GSUVMRCMjIw==\"
        }
      }
    }
  },
  {
    \"apiVersion\": \"v1\",
    \"kind\": \"Pod\",
    \"metadata\": {
      \"annotations\": {
        \"helm.sh/hook\": \"test\"
      },
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"1.16.0\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"###DYNAMIC_FIELD###\"
    },
    \"spec\": {
      \"containers\": [
        {
          \"args\": [
            \"chartsnap-app1:80\"
          ],
          \"command\": [
            \"wget\"
          ],
          \"image\": \"busybox\",
          \"name\": \"wget\"
        }
      ],
      \"restartPolicy\": \"Never\"
    }
  },
  {
    \"apiVersion\": \"v1\",
    \"kind\": \"Service\",
    \"metadata\": {
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"1.16.0\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"chartsnap-app1\"
    },
    \"spec\": {
      \"ports\": [
        {
          \"name\": \"http\",
          \"port\": 80,
          \"protocol\": \"TCP\",
          \"targetPort\": \"http\"
        },
        {
          \"name\": \"https\",
          \"port\": 443,
          \"protocol\": \"TCP\",
          \"targetPort\": \"###DYNAMIC_FIELD###\"
        }
      ],
      \"selector\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/name\": \"app1\"
      },
      \"type\": \"ClusterIP\"
    }
  },
  {
    \"apiVersion\": \"v1\",
    \"automountServiceAccountToken\": true,
    \"kind\": \"ServiceAccount\",
    \"metadata\": {
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"1.16.0\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"chartsnap-app1\"
    }
  }
]
"""

['Diff DiffContextLineN is 0 should return all diff 1']
SnapShot = """
- apiVersion: v2
+ apiVersion: v1
  automountServiceAccountToken: true
  kind: ServiceAccount
  metadata:
    labels:
      app.kubernetes.io/instance: chartsnap
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: app1
      app.kubernetes.io/version: 1.16.0
      helm.sh/chart: app1-0.1.0
    name: chartsnap-app1
- ---
- apiVersion: v1
- kind: Namespace
- metadata:
-   annotations:
-     helm.sh/hook: test
-   labels:
-     app.kubernetes.io/instance: chartsnap
-     app.kubernetes.io/managed-by: Helm
-     app.kubernetes.io/name: app1
-     app.kubernetes.io/version: 1.16.0
-     helm.sh/chart: app1-0.1.0
-   name: chartsnap-app1-namespace
  ---
  apiVersion: v1
  data:
-   ca.crt: '###DYNAMIC_FIELD###'
-   tls.crt: '###DYNAMIC_FIELD###'
-   tls.key: '###DYNAMIC_FIELD###'
+   ca.crt: IyMjRFlOQU1JQ19GSUVMRCMjIw==
+   tls.crt: IyMjRFlOQU1JQ19GSUVMRCMjIw==
+   tls.key: IyMjRFlOQU1JQ19GSUVMRCMjIw==
  kind: Secret
  metadata:
    labels:
      app.kubernetes.io/instance: chartsnap
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: app1
      app.kubernetes.io/version: 1.16.0
      helm.sh/chart: app1-0.1.0
    name: app1-cert
    namespace: default
  type: kubernetes.io/tls
  ---
  apiVersion: v1
  kind: Service
  metadata:
    labels:
      app.kubernetes.io/instance: chartsnap
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: app1
      app.kubernetes.io/version: 1.16.0
      helm.sh/chart: app1-0.1.0
    name: chartsnap-app1
  spec:
    ports:
      - name: http
        port: 80
        protocol: TCP
        targetPort: http
    selector:
      app.kubernetes.io/instance: chartsnap
      app.kubernetes.io/name: app1
-   type: LoadBalancer
+   type: ClusterIP
  ---
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    labels:
      app.kubernetes.io/instance: chartsnap
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: app1
-     app.kubernetes.io/version: 1.15.0
+     app.kubernetes.io/version: 1.16.0
      helm.sh/chart: app1-0.1.0
    name: chartsnap-app1
  spec:
    replicas: 1
    selector:
      matchLabels:
        app.kubernetes.io/instance: chartsnap
        app.kubernetes.io/name: app1
    template:
      metadata:
        labels:
          app.kubernetes.io/instance: chartsnap
          app.kubernetes.io/managed-by: Helm
          app.kubernetes.io/name: app1
          app.kubernetes.io/version: 1.16.0
          helm.sh/chart: app1-0.1.0
      spec:
        containers:
          - image: nginx:1.16.0
            imagePullPolicy: IfNotPresent
            livenessProbe:
              httpGet:
                path: /
                port: http
            name: app1
            ports:
              - containerPort: 80
                name: http
                protocol: TCP
            readinessProbe:
              httpGet:
                path: /
                port: http
            resources: {}
            securityContext: {}
        securityContext: {}
        serviceAccountName: chartsnap-app1
+ ---
+ apiVersion: networking.k8s.io/v1
+ kind: Ingress
+   annotations:
+     cert-manager.io/cluster-issuer: nameOfClusterIssuer
+   labels:
+     app.kubernetes.io/instance: chartsnap
+     app.kubernetes.io/managed-by: Helm
+     app.kubernetes.io/name: app1
+     app.kubernetes.io/version: 1.16.0
+     helm.sh/chart: app1-0.1.0
+   name: chartsnap-app1
+   ingressClassName: nginx
+   rules:
+     - host: chart-example.local
+       http:
+         paths:
+           - backend:
+               service:
+                 name: chartsnap-app1
+                 port:
+                   number: 80
+             path: /
+             pathType: ImplementationSpecific
+   tls:
+     - hosts:
+         - chart-example.local
+       secretName: chart-example-tls
  ---
  apiVersion: v1
  kind: Pod
  metadata:
    annotations:
      helm.sh/hook: test
    labels:
      app.kubernetes.io/instance: chartsnap
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: app1
      app.kubernetes.io/version: 1.16.0
      helm.sh/chart: app1-0.1.0
    name: chartsnap-app1-test-connection
  spec:
    containers:
      - args:
          - chartsnap-app1:80
        command:
          - wget
        image: busybox
        name: wget
    restartPolicy: Never
- a
+ 
"""

['Diff DiffContextLineN is 3 should return the extracted diff with previous/next 3 lines 1']
SnapShot = """
@@ KIND=ServiceAccount NAME=chartsnap-app1 LINE=1
- apiVersion: v2
+ apiVersion: v1
  automountServiceAccountToken: true
  kind: ServiceAccount
  metadata:

@@ KIND=Namespace NAME=chartsnap-app1-namespace LINE=12
- ---
- apiVersion: v1
- kind: Namespace
- metadata:
-   annotations:
-     helm.sh/hook: test
-   labels:
-     app.kubernetes.io/instance: chartsnap
-     app.kubernetes.io/managed-by: Helm
-     app.kubernetes.io/name: app1
-     app.kubernetes.io/version: 1.16.0
-     helm.sh/chart: app1-0.1.0
-   name: chartsnap-app1-namespace

@@ KIND=Secret NAME=app1-cert LINE=28
  apiVersion: v1
  data:
-   ca.crt: '###DYNAMIC_FIELD###'
-   tls.crt: '###DYNAMIC_FIELD###'
-   tls.key: '###DYNAMIC_FIELD###'
+   ca.crt: IyMjRFlOQU1JQ19GSUVMRCMjIw==
+   tls.crt: IyMjRFlOQU1JQ19GSUVMRCMjIw==
+   tls.key: IyMjRFlOQU1JQ19GSUVMRCMjIw==
  kind: Secret
  metadata:
    labels:

@@ KIND=Service NAME=chartsnap-app1 LINE=62
    selector:
      app.kubernetes.io/instance: chartsnap
      app.kubernetes.io/name: app1
-   type: LoadBalancer
+   type: ClusterIP

@@ KIND=Deployment NAME=chartsnap-app1 LINE=71
      app.kubernetes.io/instance: chartsnap
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: app1
-     app.kubernetes.io/version: 1.15.0
+     app.kubernetes.io/version: 1.16.0
      helm.sh/chart: app1-0.1.0
    name: chartsnap-app1
  spec:

@@ KIND=Ingress NAME= LINE=108
+ ---
+ apiVersion: networking.k8s.io/v1
+ kind: Ingress
+   annotations:
+     cert-manager.io/cluster-issuer: nameOfClusterIssuer
+   labels:
+     app.kubernetes.io/instance: chartsnap
+     app.kubernetes.io/managed-by: Helm
+     app.kubernetes.io/name: app1
+     app.kubernetes.io/version: 1.16.0
+     helm.sh/chart: app1-0.1.0
+   name: chartsnap-app1
+   ingressClassName: nginx
+   rules:
+     - host: chart-example.local
+       http:
+         paths:
+           - backend:
+               service:
+                 name: chartsnap-app1
+                 port:
+                   number: 80
+             path: /
+             pathType: ImplementationSpecific
+   tls:
+     - hosts:
+         - chart-example.local
+       secretName: chart-example-tls

@@ KIND=Pod NAME=chartsnap-app1-test-connection LINE=131
        image: busybox
        name: wget
    restartPolicy: Never
- a
+ 
"""

['Unknown OK report unknown as warning 1']
SnapShot = """
failed to recognize a resource in stdout/stderr of helm template command output. snapshot it as Unknown: 
---
some: raw data
raw:
  data: here
---"""
